File: sdk/javascript/src/modules/rooms.js
/**
 * Rooms module for ResCanvas SDK
 * 
 * Handles room management, drawing operations, and collaboration features.
 */

class RoomsClient {
  constructor(client) {
    this.client = client;
  }

  /**
   * Create a new drawing room
   * 
   * @param {Object} roomData - Room configuration
   * @param {string} roomData.name - Room name (1-256 chars, required)
   * @param {string} roomData.type - Room type: "public", "private", or "secure" (required)
   * @param {string} [roomData.description] - Room description (max 500 chars)
   * @returns {Promise<{room: object}>} Created room object
   */
  async create({ name, type, description }) {
    return await this.client._request('/rooms', {
      method: 'POST',
      body: JSON.stringify({ name, type, description })
    });
  }

  /**
   * List rooms accessible to authenticated user
   * 
   * @param {Object} [options={}] - List options
   * @param {boolean} [options.includeArchived] - Include archived rooms
   * @param {string} [options.sortBy] - Sort field: "createdAt", "updatedAt", "name"
   * @param {string} [options.order] - Sort order: "asc" or "desc"
   * @param {number} [options.page] - Page number for pagination
   * @param {number} [options.per_page] - Items per page
   * @param {string} [options.type] - Filter by room type
   * @returns {Promise<{rooms: array, pagination: object}>} Rooms list with pagination
   */
  async list(options = {}) {
    const params = new URLSearchParams();
    if (options.includeArchived) params.append('includeArchived', 'true');
    if (options.sortBy) params.append('sortBy', options.sortBy);
    if (options.order) params.append('order', options.order);
    if (options.page) params.append('page', options.page.toString());
    if (options.per_page) params.append('per_page', options.per_page.toString());
    if (options.type) params.append('type', options.type);

    const queryString = params.toString();
    const path = queryString ? `/rooms?${queryString}` : '/rooms';

    return await this.client._request(path, {
      method: 'GET'
    });
  }

  /**
   * Get details for a specific room
   * 
   * @param {string} roomId - Room ID
   * @returns {Promise<{room: object}>} Room details
   */
  async get(roomId) {
    return await this.client._request(`/rooms/${roomId}`, {
      method: 'GET'
    });
  }

  /**
   * Update room settings
   * 
   * @param {string} roomId - Room ID
   * @param {Object} updates - Fields to update
   * @param {string} [updates.name] - New room name
   * @param {string} [updates.description] - New description
   * @param {boolean} [updates.archived] - Archive status
   * @returns {Promise<{room: object}>} Updated room object
   */
  async update(roomId, updates) {
    return await this.client._request(`/rooms/${roomId}`, {
      method: 'PATCH',
      body: JSON.stringify(updates)
    });
  }

  /**
   * Delete a room (owner only)
   * 
   * @param {string} roomId - Room ID
   * @returns {Promise<{status: string}>} Deletion status
   */
  async delete(roomId) {
    return await this.client._request(`/rooms/${roomId}`, {
      method: 'DELETE'
    });
  }

  /**
   * Share room with multiple users
   * 
   * @param {string} roomId - Room ID
   * @param {Array<{username: string, role: string}>} users - Users to share with
   * @returns {Promise<{status: string, results: array}>} Share results
   */
  async share(roomId, users) {
    return await this.client._request(`/rooms/${roomId}/share`, {
      method: 'POST',
      body: JSON.stringify({ users })
    });
  }

  /**
   * Get list of room members with their roles
   * 
   * @param {string} roomId - Room ID
   * @returns {Promise<{members: array}>} Room members
   */
  async getMembers(roomId) {
    return await this.client._request(`/rooms/${roomId}/members`, {
      method: 'GET'
    });
  }

  /**
   * Update member permissions
   * 
   * @param {string} roomId - Room ID
   * @param {string} memberId - Member ID to update
   * @param {string} role - New role: "owner", "editor", or "viewer"
   * @returns {Promise<{status: string}>} Update status
   */
  async updatePermissions(roomId, memberId, role) {
    return await this.client._request(`/rooms/${roomId}/permissions`, {
      method: 'PATCH',
      body: JSON.stringify({ memberId, role })
    });
  }

  /**
   * Get drawing strokes for a room
   * 
   * @param {string} roomId - Room ID
   * @param {Object} [options={}] - Query options
   * @param {number} [options.since] - Get strokes after this timestamp
   * @param {number} [options.until] - Get strokes before this timestamp
   * @returns {Promise<{strokes: array}>} Room strokes
   */
  async getStrokes(roomId, options = {}) {
    const params = new URLSearchParams();
    if (options.since) params.append('since', options.since.toString());
    if (options.until) params.append('until', options.until.toString());

    const queryString = params.toString();
    const path = queryString ? `/rooms/${roomId}/strokes?${queryString}` : `/rooms/${roomId}/strokes`;

    return await this.client._request(path, {
      method: 'GET'
    });
  }

  /**
   * Submit a new drawing stroke to a room
   * 
   * @param {string} roomId - Room ID
   * @param {Object} stroke - Stroke data
   * @param {Array<{x: number, y: number}>} stroke.pathData - Array of points
   * @param {string} stroke.color - Stroke color (hex format, e.g., "#000000")
   * @param {number} stroke.lineWidth - Line width in pixels
   * @param {string} [stroke.user] - Username (optional, server can infer)
   * @param {string} [stroke.tool] - Drawing tool name
   * @param {string} [stroke.signature] - Signature for secure rooms
   * @param {string} [stroke.signerPubKey] - Signer public key for secure rooms
   * @returns {Promise<{status: string, stroke: object}>} Submission result
   */
  async addStroke(roomId, stroke) {
    return await this.client._request(`/rooms/${roomId}/strokes`, {
      method: 'POST',
      body: JSON.stringify(stroke)
    });
  }

  /**
   * Undo the last stroke submitted by current user
   * 
   * @param {string} roomId - Room ID
   * @returns {Promise<{status: string, undone: object}>} Undo result
   */
  async undo(roomId) {
    return await this.client._request(`/rooms/${roomId}/undo`, {
      method: 'POST'
    });
  }

  /**
   * Redo a previously undone stroke
   * 
   * @param {string} roomId - Room ID
   * @returns {Promise<{status: string, redone: object}>} Redo result
   */
  async redo(roomId) {
    return await this.client._request(`/rooms/${roomId}/redo`, {
      method: 'POST'
    });
  }

  /**
   * Clear all strokes from room canvas
   * 
   * @param {string} roomId - Room ID
   * @returns {Promise<{status: string, clearedAt: number}>} Clear result
   */
  async clear(roomId) {
    return await this.client._request(`/rooms/${roomId}/clear`, {
      method: 'POST'
    });
  }

  /**
   * Get undo/redo status for current user in room
   * 
   * @param {string} roomId - Room ID
   * @returns {Promise<{canUndo: boolean, canRedo: boolean}>} Undo/redo availability
   */
  async getUndoRedoStatus(roomId) {
    return await this.client._request(`/rooms/${roomId}/undo-redo-status`, {
      method: 'GET'
    });
  }

  /**
   * Reset undo/redo stacks for current user
   * 
   * @param {string} roomId - Room ID
   * @returns {Promise<{status: string}>} Reset status
   */
  async resetStacks(roomId) {
    return await this.client._request(`/rooms/${roomId}/reset-stacks`, {
      method: 'POST'
    });
  }

  /**
   * Transfer room ownership to another user
   * 
   * @param {string} roomId - Room ID
   * @param {string} newOwnerId - New owner's user ID
   * @returns {Promise<{status: string}>} Transfer status
   */
  async transferOwnership(roomId, newOwnerId) {
    return await this.client._request(`/rooms/${roomId}/transfer`, {
      method: 'POST',
      body: JSON.stringify({ newOwnerId })
    });
  }

  /**
   * Leave a room (removes membership)
   * 
   * @param {string} roomId - Room ID
   * @returns {Promise<{status: string}>} Leave status
   */
  async leave(roomId) {
    return await this.client._request(`/rooms/${roomId}/leave`, {
      method: 'POST'
    });
  }

  /**
   * Invite a user to join a room
   * 
   * @param {string} roomId - Room ID
   * @param {string} username - Username to invite
   * @param {string} [role='editor'] - Role to assign: "owner", "editor", or "viewer"
   * @returns {Promise<{status: string, inviteId: string}>} Invitation result
   */
  async invite(roomId, username, role = 'editor') {
    return await this.client._request(`/rooms/${roomId}/invite`, {
      method: 'POST',
      body: JSON.stringify({ username, role })
    });
  }

  /**
   * Search for rooms by name (autocomplete)
   * 
   * @param {string} query - Search query
   * @returns {Promise<{rooms: array}>} Matching rooms
   */
  async suggest(query) {
    return await this.client._request(`/rooms/suggest?q=${encodeURIComponent(query)}`, {
      method: 'GET'
    });
  }
}

export default RoomsClient;

File: backend/api_v1/rooms.py
"""
API v1: Room Management Endpoints

Versioned room API for external applications.
All routes are prefixed with /api/v1/rooms

Endpoints:
- POST /api/v1/rooms - Create new room
- GET /api/v1/rooms - List accessible rooms
- GET /api/v1/rooms/<id> - Get room details
- PATCH /api/v1/rooms/<id> - Update room
- DELETE /api/v1/rooms/<id> - Delete room
- POST /api/v1/rooms/<id>/share - Share room with users
- GET /api/v1/rooms/<id>/members - Get room members
- PATCH /api/v1/rooms/<id>/permissions - Update member permissions
- POST /api/v1/rooms/<id>/transfer - Transfer ownership
- POST /api/v1/rooms/<id>/leave - Leave room
- GET /api/v1/rooms/<id>/strokes - Get drawing strokes
- POST /api/v1/rooms/<id>/strokes - Submit new stroke
- POST /api/v1/rooms/<id>/undo - Undo last stroke
- POST /api/v1/rooms/<id>/redo - Redo undone stroke
- POST /api/v1/rooms/<id>/clear - Clear all strokes
- GET /api/v1/rooms/<id>/undo-redo-status - Get undo/redo status
- POST /api/v1/rooms/<id>/reset-stacks - Reset undo/redo stacks
- POST /api/v1/rooms/<id>/invite - Invite user to room
- GET /api/v1/rooms/suggest - Suggest rooms (autocomplete)
"""

from flask import Blueprint
from routes.rooms import (
    create_room,
    list_rooms,
    get_room_details,
    update_room,
    delete_room,
    share_room,
    get_room_members,
    update_permissions,
    transfer_ownership,
    leave_room,
    get_strokes,
    post_stroke,
    room_undo,
    room_redo,
    room_clear,
    get_undo_redo_status,
    reset_my_stacks,
    invite_user,
    suggest_rooms,
    admin_fill_wrapped_key
)

# Create v1 rooms blueprint with /api/v1/rooms prefix
rooms_v1_bp = Blueprint('rooms_v1', __name__, url_prefix='/api/v1/rooms')

# Room CRUD
rooms_v1_bp.add_url_rule('', 'create_room', create_room, methods=['POST'])
rooms_v1_bp.add_url_rule('', 'list_rooms', list_rooms, methods=['GET'])
rooms_v1_bp.add_url_rule('/<roomId>', 'get_room_details', get_room_details, methods=['GET'])
rooms_v1_bp.add_url_rule('/<roomId>', 'update_room', update_room, methods=['PATCH'])
rooms_v1_bp.add_url_rule('/<roomId>', 'delete_room', delete_room, methods=['DELETE'])

# Room sharing and members
rooms_v1_bp.add_url_rule('/<roomId>/share', 'share_room', share_room, methods=['POST'])
rooms_v1_bp.add_url_rule('/<roomId>/members', 'get_room_members', get_room_members, methods=['GET'])
rooms_v1_bp.add_url_rule('/<roomId>/permissions', 'update_permissions', update_permissions, methods=['PATCH'])
rooms_v1_bp.add_url_rule('/<roomId>/transfer', 'transfer_ownership', transfer_ownership, methods=['POST'])
rooms_v1_bp.add_url_rule('/<roomId>/leave', 'leave_room', leave_room, methods=['POST'])
rooms_v1_bp.add_url_rule('/<roomId>/invite', 'invite_user', invite_user, methods=['POST'])

# Drawing data
rooms_v1_bp.add_url_rule('/<roomId>/strokes', 'get_strokes', get_strokes, methods=['GET'])
rooms_v1_bp.add_url_rule('/<roomId>/strokes', 'post_stroke', post_stroke, methods=['POST'])

# Undo/Redo operations
rooms_v1_bp.add_url_rule('/<roomId>/undo', 'room_undo', room_undo, methods=['POST'])
rooms_v1_bp.add_url_rule('/<roomId>/redo', 'room_redo', room_redo, methods=['POST'])
rooms_v1_bp.add_url_rule('/<roomId>/clear', 'room_clear', room_clear, methods=['POST'])
rooms_v1_bp.add_url_rule('/<roomId>/undo-redo-status', 'get_undo_redo_status', get_undo_redo_status, methods=['GET'])
rooms_v1_bp.add_url_rule('/<roomId>/reset-stacks', 'reset_my_stacks', reset_my_stacks, methods=['POST'])

# Utilities
rooms_v1_bp.add_url_rule('/suggest', 'suggest_rooms', suggest_rooms, methods=['GET'])

# Admin endpoints (internal use only, but exposed for compatibility)
rooms_v1_bp.add_url_rule('/<roomId>/admin/fill-wrapped-key', 'admin_fill_wrapped_key', admin_fill_wrapped_key, methods=['POST'])
